stan/math/rev/functor/apply_scalar_binary.hpp

```diff
@@ -0,0 +1,248 @@
+#ifndef STAN_MATH_REV_FUNCTOR_APPLY_SCALAR_BINARY_HPP
+#define STAN_MATH_REV_FUNCTOR_APPLY_SCALAR_BINARY_HPP
+
+#include <stan/math/prim/fun/as_column_vector_or_scalar.hpp>
+#include <stan/math/rev/meta.hpp>
+#include <stan/math/prim/err/check_matching_dims.hpp>
+#include <stan/math/prim/err/check_matching_sizes.hpp>
+#include <stan/math/prim/fun/num_elements.hpp>
+#include <vector>
+
+namespace stan {
+namespace math {
+
+/**
+ * Specialisation for use with combinations of
+ * `Eigen::Matrix` and `var_value<Eigen::Matrix>` inputs.
+ * Eigen's binaryExpr framework is used for more efficient indexing of both row-
+ * and column-major inputs  without separate loops.
```

Comment:
Can drop this since `binaryExpr` isn't used with the `varmat` types

Reply:
Oh actually I had to have this for things like `f(std::vector<var<Matrix>>, std::vector<var<Matrix>>)`

Reply:
Yeah but `binaryExpr` only gets used when both inputs are treated as containers of scalars, whereas the `var<Matrix>` are treated as containers in themselves.

So `f(std::vector<var<Matrix>>, std::vector<var<Matrix>>)` evaluates to a loop over the `std::vectors` calling the function on the `var<Matrix>` inputs directly